gordon_janitor_gcp/gpubsub_client.py

```diff
@@ -0,0 +1,224 @@
+    logic. No other use cases are expected.
+
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    config = {
+        'keyfile': '/path/to/keyfile.json',
+        'project': 'a-dns-project'
+    }
+    changes_channel = asyncio.Queue()
+
+    publisher = GooglePubsubPublisher(
+        config, changes_channel)
+
+    loop = asyncio.get_event_loop()
+    try:
+        loop.run_until_complete(publisher.start())
+    finally:
+        loop.close()
+"""
+
+import asyncio
+import datetime
+import json
+import logging
+import os
+
+from asyncio_extras import threads
+from google.api_core import exceptions as google_exceptions
+from google.cloud import pubsub
+
+from gordon_janitor_gcp import auth
+from gordon_janitor_gcp import exceptions
+
+
+class GooglePubsubPublisher:
+    """Client to publish change messages to Google Pub/Sub.
+
+    Args:
+        config (dict): Google Cloud Pub/Sub-related configuration.
+        changes_channel (asyncio.Queue): queue to publish message to
+            make corrections to Cloud DNS.
+
+    Raises:
+        :py:exc:`gordon_janitor_gcp.exceptions.GCPGordonJanitorError`
+            if there are issues creating a Pub/Sub topic. This is not
+            raised if the topic already exists.
+
+        :py:exc:`gordon_janitor_gcp.exceptions.GCPConfigError`
+            when required configuration is not set. See :doc:`config`
+            for what is required.
+    """
+
+    def __init__(self, config, changes_channel=None, **kw):
+        self.config = config
+        self.changes_channel = changes_channel
+        self.cleanup_timeout = config.get('cleanup_timeout', 60)
+        self.topic = None  # set in _init
+        self.publisher = None  # set in _init
+        self._messages = set()
+        self._init()
+
+    def _init(self):
+        # a publisher client can't be made with credentials if a channel
+        # is already made/provided, which is what happens when the
+        # emulator is running ಠ_ಠ
+        # See (https://github.com/GoogleCloudPlatform/
+        #      google-cloud-python/pull/4839)
+        auth_client = None
+        if not os.environ.get('PUBSUB_EMULATOR_HOST'):
+            auth_client = self._init_auth()
+
+        self.publisher = self._init_publisher(auth_client)
+
+    def _init_auth(self):
+        # TODO (lynn): keyfile won't be required once we support other
+        #              auth methods
+        try:
+            keyfile = self.config['keyfile']
+        except KeyError:
+            msg = ('The path to a Service Account JSON keyfile is required '
+                   'to authenticate for Google Cloud Pub/Sub.')
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+
+        scopes = self.config.get('scopes')
+        return auth.GoogleAuthClient(keyfile=keyfile, scopes=scopes)
+
+    def _init_publisher(self, auth_client):
+        try:
+            project = self.config['project']
+        except KeyError:
```

Comment:
Hm, where would the plugin define the required config keys?

Reply:
Also FTR the reconciler that's already in `develop` has the same behavior. If we can figure out a way to do this more elegantly, it may be best to do it in a separate PR.